### PR TITLE
Improve regex derivative construction

### PR DESCRIPTION
**Pull request by veanes**
The initial trigger for this PR was to fix issue https://github.com/dotnet/runtime/issues/60645.
There were several root causes behind the issue that surfaced through that stress test that are being fixed by this PR.

1. The derivative computation was simplified to a more transparent algebraic recursive formulation that captures the backtracking semantics by maintaining correct order of alternatives when `OrderedOr` is constructed. This needed a separate flag `IsHighPriorityNullable` to be maintained and a method `PruneLowerPriorityThanNullability` that extracts a backtracking derivative from a full derivative when transitions are computed in Backtracking-semantics-preserving mode. This implied a reduction of several quadratic behaviors that were caused by the earlier explicit expansion of transitions. Also caching became both easier to maintain and understand in this setting.
2. The mechanism of computing _Effects_ was reformulated to use the updated derivative computation. This affected how the effects are propagated in the symbolic regex node AST and thus also affected how some of the simplification rules are triggered. Some work is remaining here to improve the way those rules are written, essentially to try to capture the algebraic laws more elegantly and to better separate the concern of checking if a _subsumption_ holds while also simplifying the regex accordingly to avoid double work.
3. More subsumption rules were added to reduce common cases of expansion that in some cases triggered quadratic (in the size of the regex) behavior and slow performance in the concatenation stress test in the 3rd phase when subcaptures are computed. The main culprit in these tests is deep nesting of sub-captures in combination with nullable patterns such as `(a?(a?(a?(a?....))))` that create a lot of possible paths in the underlying NFA.
4. Concatenation nodes are no longer normalized into right-associative form. This avoided a quadratic behavior in the length of the regex in certain cases, when the right-associative form was being recomputed every time a concatenation happened, e.g. avoiding turning `(a(b(cd))))(e(fg))` into `(a(b(c(d(e(fg))))))`

Concretely, a typical subsumption rule (avoiding alternation) is that `Y|XY` reduces to `X??Y` if `X` is a nullable regex.
This PR contains commits from me as well as @olsaarik.

### CODE REVIEW

**Comment by olsaarik** [comment]
I didn't finish commenting everything I should. Will continue tomorrow morning.

**Comment by olsaarik** [comment]
I think the state of comments for all the new code is in a relatively good place now. Feel free to review!

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:501]: In addition to the comments saying what's being done, it'd be _really_ helpful if they said why it's being done, e.g. what happens if we don't do this?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:528]: Can you add comments somewhere defining "effect", what it means to create one, how it relates to other nodes, etc?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:539]:
```suggestion
            if (effectNode == builder.Epsilon)
                return node;

            if (node == builder._nothing)
                return builder._nothing;

            if (node._kind == SymbolicRegexNodeKind.Effect)
            {
                Debug.Assert(node._left is not null && node._right is not null);
                return CreateEffect(builder, node._left, CreateConcat(builder, effectNode, node._right));
            }

            return Create(builder, SymbolicRegexNodeKind.Effect, node, effectNode, -1, -1, default, null, SymbolicRegexInfo.Effect(node._info));
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:737]: Love this comment.  More like this would be very helpful.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:752]:
```suggestion

            // Nothing has an empty language, which is subsumed by anything
            if (other == _builder._nothing)
                return true;

            // Early exit if we've gone too deep
            if (depth >= SubsumptionCheckDepthLimit)
                return false;
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:770]:
```suggestion

            // Cache and return the result if any rule applied
            if (subsumes.HasValue)
            {
                return (_builder._subsumptionCache[(this, other)] = subsumes.Value);
            }

            // Assume false if no rule applied
            return false;
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:847]:
```suggestion

                        Debug.Assert(node._left is not null && node._right is not null);
                        if (node._left != prefix._left)
                            return false;

                        node = node._right;
                        prefix = prefix._right;
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:858]:
```suggestion

                    // Handle the final element
                    if (node._kind != SymbolicRegexNodeKind.Concat)
                        return false;

                    Debug.Assert(node._left is not null && node._right is not null);
                    if (node._left == prefix)
                    {
                        tail = node._right;
                        return true;
                    }

                    return false;
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:912]:
```suggestion

                // If there are any accumulated effects we don't know how to handle them here.
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:920]:
```suggestion

                if (TryFoldAlternation(left._left, right, out SymbolicRegexNode<TSet>? innerResult, rightEffects))
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:987]:
```suggestion

                // Return false if we failed to find a split
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1383]: What does it mean for "XY" to be forgotten?  There is no "XY" in the original expression.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1376]: I'm not clear on what pruning means in this context.  We still need to keep the branch, it just doesn't have priority, right?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1429]: Can you add some examples of what it means to apply effects on transitions and also interpret effects?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1570]: I assume _right _is_ null here?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1383]: Essentially, (X|Y)Z is equivalent to (XZ|YZ) here the branch YZ is forgotten (sorry, XY was a typo, will fix), 
will get beck to addressing the comments after a meeting I have, will go through them all today.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1376]: In a situation (X|Y) where X is nullable (in the given context), Y is never used, thus omitted (forgotten).
Sort of like taking the higher-priority branch in backtracking that is known to lead to a match at which point the other branches become irrelevant and are no  longer followed.

**Comment by veanes** [comment]
I'm planning to go through all the comments, want to do it now, ... but have a couple of meetings in the way. Back soon.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1570]: Ah no, _right is guaranteed to be not null here (all Effect nodes do have a right child), but it's just not needed. Would asserting _right is not null be clearer?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs]: Added a comment with rationale, also pasting here since it's interesting:
```
// The following dictionaries are used as caches for operations that recurse over the structure of SymbolicRegexNode.
// These operations are called potentially on every step of the matching process, and they may do linear work in the
// of the pattern in each call. Thus, caching is necessary to avoid a quadratic worst-case over multiple steps of
// matching when simplification rules fail to eliminate the portions being walked over.
```
Currently with the stress tests we have removing these caches wouldn't be a problem. However, we don't have a proof that our current simplification rules always avoid these worst-cases. There might be hope for removing some of these caches with future work on the theory side.

Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs]: @veanes also tagging you since you might be interested in seeing the comment I added.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs]: @veanes I'll leave this to you.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:530]: I added a lengthy comment in SymbolicRegexNodeKind.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1576]: I added the assertion.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1431]: Done!

**Comment by olsaarik** [comment]
Outerloop tests look good; the stress tests are indeed passing now. The only regex thing popping up now is `System.Text.RegularExpressions.Tests.RegexIgnoreCaseTests.Unicode_IgnoreCase_Tests`, which seems to be failing for all engines.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs]: I fully agree with this comment. But I want to point out that `_derivativeCache` seems absolutely essential. The caches related to _pruning_ are perhaps avoidable (as you say -- with improved understanding on the theory side).

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1402]: Fixed, updated the comment to:
//in a concatenation (X|Y)Z priority is given to XZ when X is nullable
//observe that (X|Y)Z is equivalent to (XZ|YZ) and the branch YZ must be ignored
//when X is nullable (observe that XZ is nullable because this=(X|Y)Z is nullable)

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs]: Also changing the method name to `IsHighPriorityNullableFor`. 
(the name `IsHighPriorityNullableForLeftmostBranch` is definitely misleading).

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1395]: updated the comment in the code accordingly:
```
// In a alternation (X|Y) where X is nullable (in the given context), Y must be eliminated.
// Thus, taking the higher-priority branch in backtracking that is known to lead to a match
// at which point the other branches become irrelevant and must no longer be used.
```

**Comment by veanes** [comment]
All comments seem to have been addressed at this point. I am sure there will be follow-up improvements going forward (e.g. more and improved subsumption checks, triggered by more evaluations and tests). But I think these belong to future PRs.

**Comment by joperezr** [comment]
> erloop tests look good; the stress tests are indeed passing now. The only regex thing popping up now is `System.Text.RegularExpressions.Tests.RegexIgnoreCaseTests.Unicode_IgnoreCase_Tests`, which seems to be failing for all engines.

Yes, that one is on me. I just opened https://github.com/dotnet/runtime/pull/69692 to disable those tests for now.

**Review by stephentoub — APPROVED** [review]
LGTM.  There are some minor things we'll want to clean up further, but I'm going to merge to get it in for Preview 5, and to unblock subsequent work.

**Pull request merged**
